File: in_Java/1331 Rank Transform of an Array.java
class Solution {
    public int[] arrayRankTransform(int[] arr) {
        Set<Integer> hs = new HashSet<>();
        for(int i : arr) hs.add(i);
        int[] t = new int[hs.size()];
        int id = 0;
        for(int i : hs) t[id++] = i;
        Arrays.sort(t);
        int[] res = new int[arr.length];
        for(int i = 0; i < res.length; i++){
            res[i] = ido(t,arr[i]) + 1;
        }
        return res;
    }
    private int ido(int[] n, int x){
        int i = 0;
        int j = n.length-1;
        int m;
        while(i <= j){
            m = i + (j-i)/2;
            if(n[m] == x){
                return m;
            }else if(n[m] > x){
                j = m-1;
            }else{
                i = m+1;
            }
        }
        return -1;
    }
}

File: in_Python/1331 Rank Transform of an Array.py
class Solution(object):
    def arrayRankTransform(self, arr):
        """
        :type arr: List[int]
        :rtype: List[int]
        """
        t = set()
        for i in arr:
            if i not in t:
                t.add(i)
        tt = []
        for i in t:
            tt.append(i)
        tt.sort()
        d = {}
        for i in range(len(tt)):
            d[tt[i]] = i+1
        for i in range(len(arr)):
            arr[i] = d[arr[i]]
        return arr